commitments/op_commitment.go

```diff
@@ -0,0 +1,90 @@
+package commitments
+
+import (
+	"fmt"
+	"log"
+)
+
+type OPCommitmentType byte
+
+const (
+	// Keccak256Byte represents a commitment using Keccak256 hashing.
+	Keccak256Byte OPCommitmentType = 0
+	// DAServiceByte represents a commitment using a DA service.
+	DAServiceByte OPCommitmentType = 1
+)
+
+type OPCommitment struct {
+	keccak256Commitment []byte
+	genericCommitment   *DAServiceOPCommitment
+}
+
+var _ Commitment = (*OPCommitment)(nil)
+
+func Keccak256Commitment(value []byte) OPCommitment {
+	return OPCommitment{keccak256Commitment: value}
+}
+
+func DAServiceCommitment(value DAServiceOPCommitment) OPCommitment {
+	return OPCommitment{genericCommitment: &value}
+}
+
+func (e OPCommitment) IsKeccak256Commitment() bool {
+	return e.keccak256Commitment != nil
+}
+
+func (e OPCommitment) IsGenericCommitment() bool {
+	return e.genericCommitment != nil
+}
+
+func (e OPCommitment) MustKeccak256Value() []byte {
+	if e.keccak256Commitment != nil {
+		return e.keccak256Commitment
+	}
+	log.Panic("OPCommitment does not contain a Keccak256Commitment value")
+	return nil // This will never be reached, but is required for compilation.
+}
+
+func (e OPCommitment) MustDAServiceValue() DAServiceOPCommitment {
+	if e.genericCommitment != nil {
+		return *e.genericCommitment
+	}
+	log.Panic("OPCommitment does not contain a DAServiceCommitment value")
+	return DAServiceOPCommitment{} // This will never be reached, but is required for compilation.
+}
+
+func (e OPCommitment) Marshal() ([]byte, error) {
+	if e.IsGenericCommitment() {
```

Comment:
knit - control flow could be represented via switch

Reply:
I don't think so, you can't call conditionals like this in `case` statements.